clarity/src/vm/types/serialization.rs

```diff
@@ -287,6 +291,132 @@ macro_rules! check_match {
+            TypeSignature::UIntType => 16,
+            TypeSignature::BoolType => 0,
+            TypeSignature::SequenceType(SequenceSubtype::ListType(list_type)) => {
+                // u32 length as big-endian bytes
+                let list_length_encode = 4;
+                list_type
+                    .get_max_len()
+                    .checked_mul(list_type.get_list_item_type().max_serialized_size()?)
+                    .and_then(|x| x.checked_add(list_length_encode))
+                    .ok_or_else(|| CheckErrors::ValueTooLarge)?
+            }
+            TypeSignature::SequenceType(SequenceSubtype::BufferType(buff_length)) => {
+                // u32 length as big-endian bytes
+                let buff_length_encode = 4;
+                u32::from(buff_length)
+                    .checked_add(buff_length_encode)
+                    .ok_or_else(|| CheckErrors::ValueTooLarge)?
+            }
+            TypeSignature::SequenceType(SequenceSubtype::StringType(StringSubtype::ASCII(
+                length,
+            ))) => {
+                // u32 length as big-endian bytes
+                let str_length_encode = 4;
+                // ascii is 1-byte per character
+                u32::from(length)
+                    .checked_add(str_length_encode)
+                    .ok_or_else(|| CheckErrors::ValueTooLarge)?
+            }
+            TypeSignature::SequenceType(SequenceSubtype::StringType(StringSubtype::UTF8(
+                length,
+            ))) => {
+                // u32 length as big-endian bytes
+                let str_length_encode = 4;
+                // utf-8 is maximum 4 bytes per codepoint (which is the length)
+                u32::from(length)
+                    .checked_mul(4)
+                    .and_then(|x| x.checked_add(str_length_encode))
+                    .ok_or_else(|| CheckErrors::ValueTooLarge)?
+            }
+            TypeSignature::PrincipalType => {
+                // version byte + 20 byte hash160
+                let maximum_issuer_size = 21;
+                let contract_name_length_encode = 1;
+                // contract name maximum length is `MAX_STRING_LEN` (128), and ASCII
+                let maximum_contract_name = MAX_STRING_LEN as u32;
+                maximum_contract_name + maximum_issuer_size + contract_name_length_encode
+            }
+            TypeSignature::TupleType(tuple_type) => {
+                let type_map = tuple_type.get_type_map();
+                // u32 length as big-endian bytes
+                let tuple_length_encode: u32 = 4;
+                let mut total_size = tuple_length_encode;
+                for (key, value) in type_map.iter() {
+                    let value_size = value.max_serialized_size()?;
+                    total_size = total_size
+                        .checked_add(1) // length of key-name
+                        .and_then(|x| x.checked_add(key.len() as u32)) // ClarityName is ascii-only, so 1 byte per length
```

Comment:
We can safely assume here that `key.len()` is going to be at most `MAX_STRING_LEN` or something?

Reply:
Yes, it is at most `MAX_STRING_LEN`. However, the key-name is part of the type description for a tuple, so we can actually know the serialized length of the key-name component of a tuple.